source/gitlab/gitlab.go

```diff
@@ -0,0 +1,221 @@
+package gitlab
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	nurl "net/url"
+	"os"
+	"strconv"
+	"strings"
+)
+
+import (
+	"github.com/golang-migrate/migrate/v4/source"
+	"github.com/xanzy/go-gitlab"
+)
+
+func init() {
+	source.Register("gitlab", &Gitlab{})
+}
+
+var (
+	ErrNoUserInfo       = fmt.Errorf("no username:token provided")
+	ErrNoAccessToken    = fmt.Errorf("no access token")
+	ErrInvalidHost      = fmt.Errorf("invalid host")
+	ErrInvalidProjectID = fmt.Errorf("invalid project id")
+	ErrInvalidResponse  = fmt.Errorf("invalid response")
+)
+
+type Gitlab struct {
+	client *gitlab.Client
+	url    string
+
+	projectID   string
+	path        string
+	listOptions *gitlab.ListTreeOptions
+	getOptions  *gitlab.GetFileOptions
+	migrations  *source.Migrations
+}
+
+type Config struct {
+}
+
+func (g *Gitlab) Open(url string) (source.Driver, error) {
+	u, err := nurl.Parse(url)
+	if err != nil {
+		return nil, err
+	}
+
+	if u.User == nil {
+		return nil, ErrNoUserInfo
+	}
+
+	password, ok := u.User.Password()
+	if !ok {
+		return nil, ErrNoAccessToken
+	}
+
+	gn := &Gitlab{
+		client:     gitlab.NewClient(nil, password),
+		url:        url,
+		migrations: source.NewMigrations(),
+	}
+
+	if u.Host != "" {
+		err = gn.client.SetBaseURL("https://" + u.Host)
```

Comment:
Use [net/url](https://golang.org/pkg/net/url/#URL) to build the string

Reply:
I fixed it too.